Crashes in Heroku (devenv)

<!-- あくまでテンプレートなので必ずしもすべての項目を埋めなくてよい -->

Step 1: 環境
============

Heroku, Sakuten/devenv@d671d26
  
Step 2: 問題
============

再現のための手順:
-----------------

 再現はできない(Heroku上)

どうなってしまうか:
-------------------

  * 落ちる

```
2018-07-27T07:36:28.436496+00:00 heroku[web.1]: Starting process with command `gunicorn backend.app:app`
2018-07-27T07:36:31.202211+00:00 app[web.1]: [heroku-exec] Starting
2018-07-27T07:36:31.458840+00:00 app[web.1]: [2018-07-27 07:36:31 +0000] [4] [INFO] Starting gunicorn 19.9.0
2018-07-27T07:36:31.459512+00:00 app[web.1]: [2018-07-27 07:36:31 +0000] [4] [INFO] Listening at: http://0.0.0.0:56673 (4)
2018-07-27T07:36:31.459633+00:00 app[web.1]: [2018-07-27 07:36:31 +0000] [4] [INFO] Using worker: sync
2018-07-27T07:36:31.464108+00:00 app[web.1]: [2018-07-27 07:36:31 +0000] [37] [INFO] Booting worker with pid: 37
2018-07-27T07:36:31.517962+00:00 app[web.1]: [2018-07-27 07:36:31 +0000] [38] [INFO] Booting worker with pid: 38
2018-07-27T07:36:32.031062+00:00 app[web.1]: [2018-07-27 07:36:32 +0000] [38] [ERROR] Exception in worker process
2018-07-27T07:36:32.031081+00:00 app[web.1]: Traceback (most recent call last):
2018-07-27T07:36:32.031083+00:00 app[web.1]:   File "/app/.heroku/python/lib/python3.6/site-packages/gunicorn/arbiter.py", line 583, in spawn_worker
2018-07-27T07:36:32.031085+00:00 app[web.1]:     worker.init_process()
2018-07-27T07:36:32.031087+00:00 app[web.1]:   File "/app/.heroku/python/lib/python3.6/site-packages/gunicorn/workers/base.py", line 129, in init_process
2018-07-27T07:36:32.031088+00:00 app[web.1]:     self.load_wsgi()
2018-07-27T07:36:32.031090+00:00 app[web.1]:   File "/app/.heroku/python/lib/python3.6/site-packages/gunicorn/workers/base.py", line 138, in load_wsgi
2018-07-27T07:36:32.031092+00:00 app[web.1]:     self.wsgi = self.app.wsgi()
2018-07-27T07:36:32.031093+00:00 app[web.1]:   File "/app/.heroku/python/lib/python3.6/site-packages/gunicorn/app/base.py", line 67, in wsgi
2018-07-27T07:36:32.031095+00:00 app[web.1]:     self.callable = self.load()
2018-07-27T07:36:32.031096+00:00 app[web.1]:   File "/app/.heroku/python/lib/python3.6/site-packages/gunicorn/app/wsgiapp.py", line 52, in load
2018-07-27T07:36:32.031098+00:00 app[web.1]:     return self.load_wsgiapp()
2018-07-27T07:36:32.031099+00:00 app[web.1]:   File "/app/.heroku/python/lib/python3.6/site-packages/gunicorn/app/wsgiapp.py", line 41, in load_wsgiapp
2018-07-27T07:36:32.031101+00:00 app[web.1]:     return util.import_app(self.app_uri)
2018-07-27T07:36:32.031103+00:00 app[web.1]:   File "/app/.heroku/python/lib/python3.6/site-packages/gunicorn/util.py", line 350, in import_app
2018-07-27T07:36:32.031104+00:00 app[web.1]:     __import__(module)
2018-07-27T07:36:32.031106+00:00 app[web.1]:   File "/app/backend/app.py", line 1, in <module>
2018-07-27T07:36:32.031107+00:00 app[web.1]:     from api.app import create_app, initdb, generate
2018-07-27T07:36:32.031109+00:00 app[web.1]:   File "/app/backend/api/app.py", line 4, in <module>
2018-07-27T07:36:32.031111+00:00 app[web.1]:     from .routes import auth, api
2018-07-27T07:36:32.031112+00:00 app[web.1]:   File "/app/backend/api/routes/auth.py", line 6, in <module>
2018-07-27T07:36:32.031114+00:00 app[web.1]:     from api.swagger import spec
2018-07-27T07:36:32.031115+00:00 app[web.1]:   File "/app/backend/api/swagger.py", line 6, in <module>
2018-07-27T07:36:32.031117+00:00 app[web.1]:     with open('api/spec/template.yml') as f:
2018-07-27T07:36:32.031181+00:00 app[web.1]: FileNotFoundError: [Errno 2] No such file or directory: 'api/spec/template.yml'
2018-07-27T07:36:32.032074+00:00 app[web.1]: [2018-07-27 07:36:32 +0000] [38] [INFO] Worker exiting (pid: 38)
2018-07-27T07:36:32.138018+00:00 app[web.1]: [2018-07-27 07:36:32 +0000] [37] [ERROR] Exception in worker process
2018-07-27T07:36:32.138022+00:00 app[web.1]: Traceback (most recent call last):
2018-07-27T07:36:32.138024+00:00 app[web.1]:   File "/app/.heroku/python/lib/python3.6/site-packages/gunicorn/arbiter.py", line 583, in spawn_worker
2018-07-27T07:36:32.138026+00:00 app[web.1]:     worker.init_process()
2018-07-27T07:36:32.138028+00:00 app[web.1]:   File "/app/.heroku/python/lib/python3.6/site-packages/gunicorn/workers/base.py", line 129, in init_process
2018-07-27T07:36:32.138030+00:00 app[web.1]:     self.load_wsgi()
2018-07-27T07:36:32.138031+00:00 app[web.1]:   File "/app/.heroku/python/lib/python3.6/site-packages/gunicorn/workers/base.py", line 138, in load_wsgi
2018-07-27T07:36:32.138033+00:00 app[web.1]:     self.wsgi = self.app.wsgi()
2018-07-27T07:36:32.138035+00:00 app[web.1]:   File "/app/.heroku/python/lib/python3.6/site-packages/gunicorn/app/base.py", line 67, in wsgi
2018-07-27T07:36:32.138036+00:00 app[web.1]:     self.callable = self.load()
2018-07-27T07:36:32.138038+00:00 app[web.1]:   File "/app/.heroku/python/lib/python3.6/site-packages/gunicorn/app/wsgiapp.py", line 52, in load
2018-07-27T07:36:32.138040+00:00 app[web.1]:     return self.load_wsgiapp()
2018-07-27T07:36:32.138041+00:00 app[web.1]:   File "/app/.heroku/python/lib/python3.6/site-packages/gunicorn/app/wsgiapp.py", line 41, in load_wsgiapp
2018-07-27T07:36:32.138043+00:00 app[web.1]:     return util.import_app(self.app_uri)
2018-07-27T07:36:32.138045+00:00 app[web.1]:   File "/app/.heroku/python/lib/python3.6/site-packages/gunicorn/util.py", line 350, in import_app
2018-07-27T07:36:32.138046+00:00 app[web.1]:     __import__(module)
2018-07-27T07:36:32.138048+00:00 app[web.1]:   File "/app/backend/app.py", line 1, in <module>
2018-07-27T07:36:32.138051+00:00 app[web.1]:     from api.app import create_app, initdb, generate
2018-07-27T07:36:32.138052+00:00 app[web.1]:   File "/app/backend/api/app.py", line 4, in <module>
2018-07-27T07:36:32.138054+00:00 app[web.1]:     from .routes import auth, api
2018-07-27T07:36:32.138056+00:00 app[web.1]:   File "/app/backend/api/routes/auth.py", line 6, in <module>
2018-07-27T07:36:32.138057+00:00 app[web.1]:     from api.swagger import spec
2018-07-27T07:36:32.138059+00:00 app[web.1]:   File "/app/backend/api/swagger.py", line 6, in <module>
2018-07-27T07:36:32.138060+00:00 app[web.1]:     with open('api/spec/template.yml') as f:
2018-07-27T07:36:32.138068+00:00 app[web.1]: FileNotFoundError: [Errno 2] No such file or directory: 'api/spec/template.yml'
2018-07-27T07:36:32.138452+00:00 app[web.1]: [2018-07-27 07:36:32 +0000] [37] [INFO] Worker exiting (pid: 37)
2018-07-27T07:36:32.197895+00:00 app[web.1]: Traceback (most recent call last):
2018-07-27T07:36:32.197905+00:00 app[web.1]:   File "/app/.heroku/python/lib/python3.6/site-packages/gunicorn/arbiter.py", line 210, in run
2018-07-27T07:36:32.198361+00:00 app[web.1]:     self.sleep()
2018-07-27T07:36:32.198386+00:00 app[web.1]:   File "/app/.heroku/python/lib/python3.6/site-packages/gunicorn/arbiter.py", line 360, in sleep
2018-07-27T07:36:32.198745+00:00 app[web.1]:     ready = select.select([self.PIPE[0]], [], [], 1.0)
2018-07-27T07:36:32.198785+00:00 app[web.1]:   File "/app/.heroku/python/lib/python3.6/site-packages/gunicorn/arbiter.py", line 245, in handle_chld
2018-07-27T07:36:32.199072+00:00 app[web.1]:     self.reap_workers()
2018-07-27T07:36:32.199112+00:00 app[web.1]:   File "/app/.heroku/python/lib/python3.6/site-packages/gunicorn/arbiter.py", line 525, in reap_workers
2018-07-27T07:36:32.199539+00:00 app[web.1]:     raise HaltServer(reason, self.WORKER_BOOT_ERROR)
2018-07-27T07:36:32.199620+00:00 app[web.1]: gunicorn.errors.HaltServer: <HaltServer 'Worker failed to boot.' 3>
2018-07-27T07:36:32.199627+00:00 app[web.1]: 
2018-07-27T07:36:32.199629+00:00 app[web.1]: During handling of the above exception, another exception occurred:
2018-07-27T07:36:32.199631+00:00 app[web.1]: 
2018-07-27T07:36:32.199635+00:00 app[web.1]: Traceback (most recent call last):
2018-07-27T07:36:32.199662+00:00 app[web.1]:   File "/app/.heroku/python/bin/gunicorn", line 11, in <module>
2018-07-27T07:36:32.199831+00:00 app[web.1]:     sys.exit(run())
2018-07-27T07:36:32.199837+00:00 app[web.1]:   File "/app/.heroku/python/lib/python3.6/site-packages/gunicorn/app/wsgiapp.py", line 61, in run
2018-07-27T07:36:32.200036+00:00 app[web.1]:     WSGIApplication("%(prog)s [OPTIONS] [APP_MODULE]").run()
2018-07-27T07:36:32.200042+00:00 app[web.1]:   File "/app/.heroku/python/lib/python3.6/site-packages/gunicorn/app/base.py", line 223, in run
2018-07-27T07:36:32.200310+00:00 app[web.1]:     super(Application, self).run()
2018-07-27T07:36:32.200314+00:00 app[web.1]:   File "/app/.heroku/python/lib/python3.6/site-packages/gunicorn/app/base.py", line 72, in run
2018-07-27T07:36:32.200512+00:00 app[web.1]:     Arbiter(self).run()
2018-07-27T07:36:32.200517+00:00 app[web.1]:   File "/app/.heroku/python/lib/python3.6/site-packages/gunicorn/arbiter.py", line 232, in run
2018-07-27T07:36:32.200777+00:00 app[web.1]:     self.halt(reason=inst.reason, exit_status=inst.exit_status)
2018-07-27T07:36:32.200783+00:00 app[web.1]:   File "/app/.heroku/python/lib/python3.6/site-packages/gunicorn/arbiter.py", line 345, in halt
2018-07-27T07:36:32.201102+00:00 app[web.1]:     self.stop()
2018-07-27T07:36:32.201149+00:00 app[web.1]:   File "/app/.heroku/python/lib/python3.6/site-packages/gunicorn/arbiter.py", line 393, in stop
2018-07-27T07:36:32.201498+00:00 app[web.1]:     time.sleep(0.1)
2018-07-27T07:36:32.201503+00:00 app[web.1]:   File "/app/.heroku/python/lib/python3.6/site-packages/gunicorn/arbiter.py", line 245, in handle_chld
2018-07-27T07:36:32.201788+00:00 app[web.1]:     self.reap_workers()
2018-07-27T07:36:32.201825+00:00 app[web.1]:   File "/app/.heroku/python/lib/python3.6/site-packages/gunicorn/arbiter.py", line 525, in reap_workers
2018-07-27T07:36:32.202253+00:00 app[web.1]:     raise HaltServer(reason, self.WORKER_BOOT_ERROR)
2018-07-27T07:36:32.202288+00:00 app[web.1]: gunicorn.errors.HaltServer: <HaltServer 'Worker failed to boot.' 3>
2018-07-27T07:36:32.288585+00:00 heroku[web.1]: State changed from starting to crashed
2018-07-27T07:36:32.274899+00:00 heroku[web.1]: Process exited with status 1
```  


どうなるべきだったのか:
-----------------------

  * 落ちない
  
関連するコード:
---------------

https://github.com/Sakuten/backend/blob/7c5fdb9af213bb188276cc1a4c6fc3f8ba492d5a/api/swagger.py#L6

--- a/api/config.py
+++ b/api/config.py
@@ -55,12 +55,6 @@ class PreviewDeploymentConfig(BaseConfig):
     SQLALCHEMY_DATABASE_URI = os.environ.get('DATABASE_URL')
     SECRET_KEY = os.environ.get('SECRET_KEY')
     ENV = 'development'
-    START_DATETIME = datetime(2018, 8, 2, 0, 0, 0, tzinfo=BaseConfig.TIMEZONE)
-    END_DATETIME = datetime(2018, 8, 2, 23, 59, 59, tzinfo=BaseConfig.TIMEZONE)
-    DRAWING_TIME_EXTENSION = timedelta(minutes=1)
-    TIMEPOINTS = [
-        (time(14,  0), time(14,  5)),
-    ]
 
 
 class DeploymentConfig(BaseConfig):

--- a/api/config.py
+++ b/api/config.py
@@ -48,6 +48,8 @@ class TestingConfig(BaseConfig):
 class PreviewDeploymentConfig(BaseConfig):
     DEBUG = True
     TESTING = False
+    ID_LIST_FILE = BaseConfig.ROOT_DIR / 'cards/test_users.json'
+    WINNERS_NUM = 3  # just small value
     # DATABASE_URL is to be set by Heroku
     # SECRET_KEY is to be set in config vars
     SQLALCHEMY_DATABASE_URI = os.environ.get('DATABASE_URL')

--- a/api/config.py
+++ b/api/config.py
@@ -55,6 +55,12 @@ class PreviewDeploymentConfig(BaseConfig):
     SQLALCHEMY_DATABASE_URI = os.environ.get('DATABASE_URL')
     SECRET_KEY = os.environ.get('SECRET_KEY')
     ENV = 'development'
+    START_DATETIME = datetime(2018, 8, 2, 0, 0, 0, tzinfo=BaseConfig.TIMEZONE)
+    END_DATETIME = datetime(2018, 8, 2, 23, 59, 59, tzinfo=BaseConfig.TIMEZONE)
+    DRAWING_TIME_EXTENSION = timedelta(minutes=1)
+    TIMEPOINTS = [
+        (time(14,  0), time(14,  5)),
+    ]
 
 
 class DeploymentConfig(BaseConfig):

--- a/api/swagger.py
+++ b/api/swagger.py
@@ -3,7 +3,7 @@ from flasgger import Swagger, swag_from
 import yaml
 import api
 
-with open('api/spec/template.yml') as f:
+with open(os.path.join(api.__path__[0], 'spec/template.yml')) as f:
     template = yaml.load(f)
 swag = Swagger(template=template)
 

